runtime/lua/vim/net.lua

```diff
@@ -0,0 +1,81 @@
+local M = {}
+
+--- Makes an HTTP GET request to the given URL.
+---
+--- This function operates in two modes:
+---   - Synchronous (blocking): If `on_response` is omitted, it waits for the
+---   request to complete and returns the result.
+---   - Asynchronous (non-blocking): If `on_response` is provided, it returns
+---   immediately and invokes the callback with the result later.
+---
+--- @param url string The URL for the request.
+--- @param opts? table Optional parameters:
+---   - `verbose` (boolean|nil): Enables verbose output.
+---   - `retry`   (integer|nil): Number of retries on transient failures (default: 3).
+---   - `outpath`  (string|nil): File path to save the response body to. If set, the `text` value in the Response Object will be `true` instead of the response body.
+--- @param on_response? fun(err?: string, content?: table|boolean) Optional callback for async execution.
+---   It is invoked with `(err, nil)` on failure or `(nil, content)` on success.
+--- @return table|boolean|nil Response Object { text: `string` | `boolean` } and `nil` on failure (sync).
+--- (Always `true` for async invocation)
+--- @return string|nil Error message on failure, or `nil` on success. (Always `nil` for async
+--- invocation)
+function M.request(url, opts, on_response)
+  vim.validate({
+    url = { url, 'string' },
+    opts = { opts, 'table', true },
+    on_response = { on_response, 'function', true },
+  })
+
+  opts = opts or {}
+  local retry = opts.retry or 3
+
+  -- Build curl command
+  local args = { 'curl' }
+  if opts.verbose then
+    table.insert(args, '--verbose')
+  else
+    vim.list_extend(args, { '--silent', '--show-error', '--fail' })
+  end
+  vim.list_extend(args, { '--location', '--retry', tostring(retry) })
+
+  if opts.outpath then
+    vim.list_extend(args, { '--output', opts.outpath })
+  end
+
+  table.insert(args, url)
+
+  local job_result = {}
+
+  local function on_exit(res)
+    local err_msg = nil
+    local response_obj = nil
+
+    if res.code ~= 0 then
+      err_msg = (res.stderr ~= '' and res.stderr)
+        or string.format('Request failed with exit code %d', res.code)
+    else
+      response_obj = {
+        text = opts.outpath and true or res.stdout,
+      }
+    end
+
+    job_result.err = err_msg
+    job_result.response = response_obj
+
+    if on_response then
+      on_response(job_result.err, job_result.response)
+    end
+  end
+
+  local job = vim.system(args, { text = true }, on_exit)
```

Comment:
Is `text=true` always valid? What happens if a `.jpg` url is requested ?

Maybe we should drop `text=true` here, and return `{body: string}` and let the caller decide how to treat `body` for now. (Note: Lua strings can have binary data.)